### PR TITLE
Add a mixed precision test and fix mixed precision errors for layers

### PR DESCRIPTION
**Pull request by mattdangerw**

### CODE REVIEW

**Comment by mattdangerw** [comment]
Ah I think this needs a keras-core release to pass. Will sync with Francois.

File: keras_nlp/layers/modeling/transformer_decoder.py [hunk 1]
```diff
@@ -218,25 +225,30 @@ def build(
...
+            name="feedforward_output_dense",
         )
         intermediate_shape = list(decoder_sequence_shape)
         intermediate_shape[-1] = self.intermediate_dim
         self._feedforward_output_dense.build(tuple(intermediate_shape))
         self._feedforward_layer_norm = keras.layers.LayerNormalization(
             epsilon=self.layer_norm_epsilon,
-            name="output_layer_norm",
+            dtype=self.dtype_policy,
+            name="feedforward_layer_norm",
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/transformer_decoder.py:245]: Just a random thought while I saw these 

With the new distribution API, we're going to need to be careful about changing layer names moving forward!

File: keras_nlp/samplers/sampler.py [hunk 2]
```diff
@@ -164,6 +164,17 @@ def body(prompt, cache, index):
         )
         return prompt
 
+    def compute_probabilities(self, logits):
+        """Compute token probabilities from logits.
+
+        This will always be done in full precision, regardless of dtype, and
+        scale by `temperature`.
+        """
+        dtype = logits.dtype
```
Comment (ianstenbit) [inline keras_nlp/samplers/sampler.py:173]: Maybe `logits_dtype`? (For consistency of style with `inputs_dtype` a few files up)

File: keras_nlp/tests/test_case.py [hunk 3]
```diff
@@ -168,6 +173,27 @@ def call(self, x):
...
+
+        if run_mixed_precision_check:
+            layer = layer_cls(**{**init_kwargs, "dtype": "mixed_float16"})
+            if isinstance(input_data, dict):
+                output_data = layer(**input_data)
+            else:
+                output_data = layer(input_data)
+            for tensor in tree.flatten(output_data):
+                dtype = standardize_dtype(tensor.dtype)
+                if "float" in dtype:
```
Comment (ianstenbit) [inline keras_nlp/tests/test_case.py:190]: Seems like `assertDType` should be good here no?

**Review by ianstenbit — APPROVED** [review]

File: keras_nlp/layers/modeling/transformer_decoder.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/transformer_decoder.py:245]: Yep! Trying to get us in a nice consistent state before we do.

**Comment by jbischof** [comment]
Is mixed precision working in Keras Core now?

**Comment by mattdangerw** [comment]
> Is mixed precision working in Keras Core now?

Yes, or at least it should be largely with the latest release. Landed the loss scaling optimizer, which was the main piece we were missing, as well as a few other fixes.

**Comment by mattdangerw** [comment]
Merging, last breakage unrelated #1251
